use regex to parse postgres version numbers

troyk opened the pull request:
got an error parsing "9.4beta2", see #53 

This pull request uses a regex and a case, the regex may be hard to read for some but overall this pr also lowers the LOC ;)  On http://www.postgresql.org/support/versioning they do not mention if a minor version is required, but in the regex I made it optional.

Thanks

ericmj commented:
We shouldn't use regexes when we can use the much faster binary matches. You can keep the current code and use `Integer.parse/1` on the patch part to strip the beta string.

troyk commented:
ok, switching to `Integer.parse/1` brings up another issue `{badmatch,[9,4]}` because `9.4beta2` has no patch.  I have no idea if pg will report 9.4.0 as 9.4, or if this is just an issue with beta version numbers.  So using your feedback I thought I'd give String.split a shot and benchmark it, turns out its the fastest.  What you think about using `String.split/2` with `Integer.parse/1`?

 Here are the results on my macbook air:

```
PgVersionToIntBench.split:     1000000   1.62 µs/op
PgVersionToIntBench.binary:    1000000   1.91 µs/op
PgVersionToIntBench.regexp:    1000000   2.72 µs/op
```

The String.split method:

```elixir
  def version_to_int_split(version) do
    parts = String.split(version,".")
    case Enum.map(parts, &:erlang.binary_to_integer/1) do
        [major, minor, patch] -> major*10_000 + minor*100 + patch
        [major, minor] -> major*10_000 + minor*100
        [major] -> major*10_000 
    end
  end
```

On an aside, I am very new to Elixir and am very interested to think more in terms of functional matching, so I appreciate the feedback on writing more performant code.   I think removing `binary_split(binary, _pattern, 0)` and `binary_split(binary, pattern, max)` makes the code more readable, but that could be all about my Elixir noobness.

ericmj commented:
Performance is not the main issue with regexes, `~r/(\d{1,2})\.?(\d{1,3})?\.?(\d{1,3})?/` is waaaay harder to understand than `String.split(foo, ".")`.

`binary_split/3` seems to exist for legacy reasons. You can use `:binary.split(version, ".", [:global])` instead.

The `String.split/2` method doesn't seem to handle `beta` strings, please use `Integer.parse/1` to handle it. And handle any number of version parts since postgres seems to be really inconsistent here.
